text/0000-native-web-component-support.md

```diff
@@ -0,0 +1,143 @@
+---
+title: Native Web Component Support
+status: DRAFTED
+created_at: 2021-10-21
+updated_at: YYYY-MM-DD
+pr: https://github.com/salesforce/lwc-rfcs/pull/58
+---
+
+# Native Web Component Support
+
+## Summary
+
+It is possible to author LWC components that can be used natively in an off-platform application by
+registering the `CustomElementConstructor` getter.
+
+```js
+import { LightningElement } from 'lwc';
+class HelloWorld extends LightningElement {}
+customElements.define('hello-world', HelloWorld.CustomElementConstructor);
+```
+
+```html
+<html>
+  <body>
+    <hello-world></hello-world>
+  </body>
+</html>
+```
+
+The usage of native web components is generally not allowed on the platform so the main use case is
+off-platform applications.
+
+The following example attempts to use native web components in an LWC application, which is not
+supported.
+
+```js
+customElements.define('hello-world', class HelloWorld {});
+```
+
+```html
+<template>
+  <hello-world></hello-world>
+</template>
+```
+
+This is because, by design, LWC relies on a custom module loader and never defers to the custom
+element registry when resolving components. The Salesforce platform is based on a multi-tenant
+architecture, and as such, it relies on a directory-based module resolution system that is
+compatible with how developers create components on the platform. This basically guarantees that no
+two parties can define the same component, allowing LWC to side-step a variety of security concerns.
+
+This proposal is about adding native web component support for off-platform users, with the
+additional goal of exploring how we might do the same for platform users. Any solution should be
+forward-compatible with [Scoped Custom Element Registries].
+
+## Motivation
+
+The long term goal of LWC is to provide a thin abstraction layer on top of native browser APIs.
+Adding support for the custom element registry for off-platform applications would be a step towards
+that goal.
+
+For platform applications, the major motivation is to facilitate platform integration and allow
+developers to migrate existing tooling and solutions without having to utilize LWC.
+
+## Detailed design
+
+### OSS restriction
+
+In order for LWC to support native web components for OSS applications, the following needs to
+happen.
+
+1. Prevent LWC components from using the same custom element as a native web component.
+2. Prevent native web components from using the same custom element as an LWC component.
+3. Remove the restriction that LWC must know about all available components before rendering.
+

```

Comment:
Also to solve:
4. Passing attributes vs properties
5. Passing non-primitive HTML types

Reply:
Looking at the test suite for [Custom Elements Everywhere](https://custom-elements-everywhere.com/), it looks like they recommend:

- Arrays/objects should be passed as properties
- Booleans/numbers/strings can be passed as either properties or attributes

In order to pass this test suite, we will also need to handle events with `names-like-this` or `NamesLikeThis`, which LWC currently doesn't support. (At least, in the declarative template syntax, which is what the test suite explicitly tests for.) I.e.:

```html
<!-- currently does not work -->
<some-element onkebab-case={foo}></some-element>
<another-element onPascalCase={foo}></another-element>
```

Reply:
> In order to pass this test suite, we will also need to handle events with `names-like-this` or `NamesLikeThis`, which LWC currently doesn't support.

For this, we will certainly need to drop `parse5` and use a custom HTML parser. This will be quite a large undertaking.